test/unit/lib/plugins/aws/deployFunction.test.js

```diff
@@ -548,3 +549,256 @@ describe('AwsDeployFunction', () => {
     });
   });
 });
+
+describe('AwsDeployFunction - runServerless', () => {
+  describe('when ecr is used', () => {
+    const awsRequestStubMap = {
+      Lambda: {
+        getFunction: {
+          Configuration: {
+            LastModified: '2020-05-20T15:34:16.494+0000',
+          },
+        },
+      },
+    };
+
+    it('should fail if `functions[].image` references image without path and uri', async () => {
```

Comment:
I think those tests should be in `provider.test.js` where we confirm custom validation of properties.

(in general I would test functionalities in test files which correspond to files in which they were implemented)

Also I guess we can use `package` command

Reply:
How would you go about ensuring that the resolution happens correctly also for `deploy function -f functionName`? While the functionality is in `provider.js`, both `deploy` and `deploy function` use it a bit differently.

Reply:
Very good point (and I think in tons of other places we forgot about packaging that's implied by `sls deploy function`.

I think we should probably ensure one sanity test (happy path) for both `sls package` (implied by `sls deploy`) and `sls deploy function`, but then test specific scenarios on just one of them.

Otherwise ideally it would be to have this functionality implemented independently of `Serverless` instance (it could take a _imageName_ and _imagePath_), and then we can do unit tests against that uitlity, and with `runServerless` just confirm happy path for `sls package` and `sls deploy function`.

Still I guess it's not possible until we have  https://github.com/serverless/serverless/issues/8445 done (and it's likely that some other things will stand in a way)

(e.g with refactor I'm working on, I'm secluding all those bricks into independent utils, that makes right unit testing easier)

Reply:
I really like that suggestion and that's what I went with - happy path cases in both `functions.test.js` and `deployFunction.test.js` and full coverage in `provider.test.js`. I agree that ideally, I would love to have image resolution implemented as independent util, but at the moment it's hard as there are direct dependencies on `provider` instance (the biggest one on `request` that you mentioned). But I know we're on a good path to better modularization :raised_hands: